<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>promise的几个关键问题</title>
</head>
<body>
  
</body>
<script>
  /* 
    1.如何改变promise的状态？
      (1) 调用resolve(value)函数：如果当前是pending就会变为resolved
      (2) 调用reject(reason)函数：如果当前是pending就会变为rejected
      (3) 抛出异常 throw：如果当前是pending就会变为rejected
  */

  /*
    2.一个promise指定多个成功/失败回调函数，都会调用么？
      当promise改变为对象状态时都会调用
  */
 
  /*
    3.改变promise状态和指定回调函数谁先谁后？
      (1) 都有可能，正常情况下是先指定回调再改变状态，但也可以先改变状态再指定回调
      (2) 如何先改状态再指定回调？
        1）在执行器中直接调用 resolve()/reject()
        2）延迟更长时间才调用 then()
      (3) 如何先指定回调再改状态？
        1）在执行器中执行异步操作
      (3) 什么时候才能得到数据？（其实就是回调函数什么时候执行）
        1）如果先指定的回调，那当状态发生改变时，回调函数就会调用，得到数据
        2）如果先改变的状态，那当指定回调时，回调函数就会调用，得到数据
  */

  /*
    4.promise.then()返回的新promise的结果状态由什么决定？
      (1) 简单表达：由then()指定的回调函数执行的结果决定
      (2) 详细表达：
        1）如果抛出异常，新promise变为rejected，reason为抛出的异常
        2）如果返回的是非promise的任意值，新promise变为resolved，value为返回的值
        3）如果返回的是另一个新promise，此promise的结果就会成为新promise的结果
  */
    let p = new Promise((resolve, reject)=>{
      resolve('OK')
    })
    p.then(value=>{
      return new Promise((resolve, reject)=>{
        resolve('Success')
      })
    }).then(value=>{
      console.log(value) // Success
    })
  /*  
    5.promise如何串连多个操作任务？
      (1) promise的then()返回一个新的promise，可以开成then()的链式调用
      (2) 通过then的链式调用串连多个同步/异步任务
  */

  /* 
    6.promise异常穿透？
    (1) 当使用promise的then链式调用时，可以在最后指定失败的回调
    (2) 前面任何操作出了异常，都会传到最后失败的回调中处理
  */
    
  let p1 = new Promise((resolve, reject)=>{
      resolve('OK')
      // reject('Err')
  })
  p1.then(value=>{
    throw '失败了！！'
    console.log(111)
  }).then(value=>{
    console.log(222)
  }).then(value=>{
    console.log(333)
  }).catch(reason=>{
    console.warn(reason)
  })

  /* 
    中断promise链？
    (1) 当使用promise的then链式调用时，在中间中断，不再调用后面的回调函数
    (2) 办法：在回调函数中返回一个pendding状态的promise对象
  */

  let p2 = new Promise((resolve, reject)=>{
      resolve('OK')
      // reject('Err')
  })
  p2.then(value=>{
    console.log(111)
    // 有且只有一种方式 返回一个pendding状态的promise对象
    return new Promise(()=>{});
  }).then(value=>{
    console.log(222)
  }).then(value=>{
    console.log(333)
  }).catch(reason=>{
    console.warn(reason)
  })
</script>
</html>